File: leetcode 20.java
//Solution1: stack T: O(n) S: O(n)
class Solution {
    public boolean isValid(String s) {
        if(s.length()%2 == 1) return false; // If the atring has a odd length, return false
        Stack<Character> stack = new Stack<>();
        
        for(char str : s.toCharArray()) {
            if(str == '(') stack.push(')');
            else if(str == '[') stack.push(']');
            else if(str == '{') stack.push('}');
            else if(stack.isEmpty() || stack.pop() != str) return false;
        }
        return stack.isEmpty();
        
    }
}

File: python version/leetcode 20.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        aList = list(s)
        stack = []
        for pa in s:
            if pa == '(':
                stack.append(')')
            elif pa == '[':
                stack.append(']')
            elif pa == '{':
                stack.append('}')
            else:
                if len(stack) == 0 or stack.pop() != pa:
                    return False
        return stack == []
                
        